### PR TITLE
fix external etcd cluster support.

### PR DESCRIPTION
**Pull request by coresolve**
With this change:
If you set `tectonic_etcd_tls_enabled`
- We will use the `https` scheme.
- The default remains true

If you set `tectonic_etcd_servers`
- We will assume we are not generating certs for the etcd cluster
- Optionally, you can provide `tectonic_etcd_ca_cert_path` if needed.

@squat @kyoto 

With this change the the ui will have to set tectonic_etcd_tls_enabled appropriately based on scheme.

### CODE REVIEW

File: platforms/aws/tectonic.tf [hunk 1]
```diff
@@ -20,7 +20,7 @@ module "etcd_certs" {
   etcd_cert_dns_names   = "${data.template_file.etcd_hostname_list.*.rendered}"
   etcd_client_cert_path = "${var.tectonic_etcd_client_cert_path}"
   etcd_client_key_path  = "${var.tectonic_etcd_client_key_path}"
-  self_signed           = "${var.tectonic_self_hosted_etcd != "" || var.tectonic_etcd_tls_enabled}"
+  self_signed           = "${var.tectonic_self_hosted_etcd != "" ? 0 : length(compact(var.tectonic_etcd_servers)) == 0 ? 0 : 1}"
```
Comment (squat) [inline platforms/aws/tectonic.tf]: Why is self_signed being turned into an integer? I think it is much easier to reason about these things as booleans (or their terraform equivalent). Furthermore, although in POSIX we treat exit code 0 as a positive, in most interpretted languages, 0 is a falsey value and 1 is truthy. In fact, in terraform we use `count = 0` to disable components. So I would ask that if we absolutely need to go the route of converting this variable to an integer that we reverse the values. Still, I would prefer to keep this a boolean.

File: platforms/metal/remote.tf [hunk 2]
```diff
@@ -1,5 +1,5 @@
 resource "null_resource" "etcd_secrets" {
-  count = "${var.tectonic_etcd_tls_enabled ? length(var.tectonic_metal_controller_domains) : 0}"
+  count = "${length(compact(var.tectonic_etcd_servers)) == 0 ? length(var.tectonic_metal_controller_domains) : 0}"
```
Comment (squat) [inline platforms/metal/remote.tf]: What if TLS is disabled? then we don't want to provision anything here.

File: config.tf [hunk 3]
```diff
@@ -176,6 +176,9 @@ variable "tectonic_etcd_servers" {
   description = <<EOF
 (optional) List of external etcd v3 servers to connect with (hostnames/IPs only).
 Needs to be set if using an external etcd cluster.
+If these are defined we will assume that we are not self signing the certs.
```
[the remark this answers is not in the document]
Comment (coresolve) [reply config.tf]: Addressed

File: config.tf [hunk 4]
```diff
@@ -176,6 +176,9 @@ variable "tectonic_etcd_servers" {
...
+If you want to provide a ca to use to verify the etcd servers. You must also set "tectonic_etcd_ca_cert_path".
+Otherwise, it is assumed that the apiserver will be able to trust the etcd certificate using /usr/share/ca-certificates on the node
```
[the remark this answers is not in the document]
Comment (coresolve) [reply config.tf]: Addressed

File: config.tf [hunk 5]
```diff
@@ -188,7 +191,7 @@ variable "tectonic_etcd_tls_enabled" {
   default = true
 
   description = <<EOF
-(optional) If set to `true`, TLS secure communication for self-provisioned etcd. will be used.
+(optional) If set to `true`, We will configure all etcd endpoints to use the "https" scheme.
```
[the remark this answers is not in the document]
Comment (coresolve) [reply config.tf]: Addressed

File: platforms/aws/tectonic.tf (hunk 1 above)
Comment (coresolve) [reply platforms/aws/tectonic.tf]: Addressed

File: platforms/metal/remote.tf (hunk 2 above)
Comment (coresolve) [reply platforms/metal/remote.tf]: Addressed

File: Documentation/variables/config.md [hunk 6]
```diff
@@ -26,12 +26,12 @@ This document gives an overview of variables used in all platforms of the Tecton
...
 | tectonic_etcd_count | The number of etcd nodes to be created. If set to zero, the count of etcd nodes will be determined automatically.<br><br>Note: This is not supported on bare metal. | string | `0` |
-| tectonic_etcd_servers | (optional) List of external etcd v3 servers to connect with (hostnames/IPs only). Needs to be set if using an external etcd cluster.<br><br>Example: `["etcd1", "etcd2", "etcd3"]` | list | `<list>` |
-| tectonic_etcd_tls_enabled | (optional) If set to `true`, TLS secure communication for self-provisioned etcd. will be used.<br><br>Note: If `tectonic_experimental` is set to `true` this variable has no effect, because the experimental self-hosted etcd always uses TLS. | string | `true` |
+| tectonic_etcd_servers | (optional) List of external etcd v3 servers to connect with (hostnames/IPs only). Needs to be set if using an external etcd cluster. If these are defined, the installer will not create self signing the certs. To provide a CA certificate to trust the etcd servers, set "tectonic_etcd_ca_cert_path".<br><br>Example: `["etcd1", "etcd2", "etcd3"]` | list | `<list>` |
```
Comment (squat) [inline Documentation/variables/config.md]: There is something funny in the second sentence. Please correct it to say:
```
Note: if this variable is defined, the installer will not create self-signed certificates for etcd.
```

**Comment by squat** [comment]
ok to test

**Comment by sudhaponnaganti** [comment]
hmm @coresolve @squat @yifan-gu  - These checks are not successful. I think we need this for 1.7.9 RC3, right?

**Comment by squat** [comment]
ok to test

**Comment by squat** [comment]
Sudha, tests succeeded once. Ran a second time and one of the tests flaked. Running once more for confidence. I think this will be good to go then we can merge #2270

**Comment by yifan-gu** [comment]
@kyoto @coresolve @squat I can't cherry-pick this one cuz there are conflicts involved, so I'd rather wait for you guys to cherry-pick it.

**Pull request merged**